projects/plugins/protect/src/js/hooks/use-onboarding/index.jsx

```diff
@@ -0,0 +1,100 @@
+import { useDispatch, useSelect } from '@wordpress/data';
+import { createContext, useCallback, useContext, useEffect, useMemo, useState } from 'react';
+import API from '../../api';
+import { STORE_ID } from '../../state/store';
+import useProtectData from '../use-protect-data';
+
+export const OnboardingContext = createContext( [] );
+export const OnboardingRenderedContext = createContext( [] );
+
+export const OnboardingRenderedContextProvider = ( { children } ) => {
+	const [ renderedSteps, setRenderedSteps ] = useState( [] );
+
+	return (
+		<OnboardingRenderedContext.Provider value={ { renderedSteps, setRenderedSteps } }>
+			{ children }
+		</OnboardingRenderedContext.Provider>
+	);
+};
+
+const useOnboarding = () => {
+	const { hasRequiredPlan } = useProtectData();
+	const { completeOnboardingSteps, fetchOnboardingProgress } = API;
+
+	const steps = useContext( OnboardingContext );
+	const { renderedSteps } = useContext( OnboardingRenderedContext );
+
+	const progress = useSelect( select => select( STORE_ID ).getOnboardingProgress() );
+	const { setOnboardingProgress } = useDispatch( STORE_ID );
+
+	/**
+	 * Current Step
+	 *
+	 * @returns {null|object}
+	 */
+	const { currentStep, currentStepCount, stepsCount } = useMemo( () => {
+		return steps.reduce(
+			( carry, step ) => {
+				if ( renderedSteps.includes( step.id ) ) {
+					carry.stepsCount++;
+					if ( ! carry.currentStep && ( progress || [] ).indexOf( step.id ) === -1 ) {
+						carry.currentStep = step;
+						carry.currentStepCount = carry.stepsCount;
+					}
+				}
+				return carry;
+			},
+			{
+				currentStep: null,
+				currentStepCount: null,
+				stepsCount: 0,
+			}
+		);
+	}, [ progress, renderedSteps, steps ] );
+
+	const completeCurrentStep = useCallback( () => {
+		if ( currentStep ) {
+			// Complete the step immediately in the UI
+			setOnboardingProgress( [ ...progress, currentStep.id ] );
+			// Save the completion in the background
+			completeOnboardingSteps( [ currentStep.id ] );
+		}
+	}, [ currentStep, setOnboardingProgress, progress, completeOnboardingSteps ] );
+
+	/**
+	 * Complete All Current Steps
+	 */
+	const completeAllCurrentSteps = useCallback( () => {
+		const stepIds = steps
+			.filter( step => {
+				// If hasRequiredPlan is true, include paid steps, else include free steps
+				return hasRequiredPlan ? step.id.startsWith( 'paid-' ) : step.id.startsWith( 'free-' );
```

Comment:
[not blocking] Continuation from my above comment, another pro for keeping "what step displays when/where" - this logic is handy, but hidden from sight in the actual view files. Moving this to the UI level will make it (IMO) easier to understand and debug in the future.

Reply:
I will look into an alternative but I believe this one might be more of a challenge than the initial instance. We need a way within the method for completing all steps to determine what steps to complete based on the current plan status so that we leave the paid flow incomplete when the free flow is completed and vice versa. We _could_ declare separate methods for completing all paid steps and one that completes all free steps?